---
title: "LMMS CLR Trajectories"
author: "Paul Nguyen"
date: "21/12/2021"
output: pdf_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(biomformat)
library(dada2) 
library(matR) 
library(ggplot2) 
library(tidyverse) 

library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(compositions) 
library(timeOmics)
library(lmms)
library(stringi)
library(purrr)

#Change this line to /LMMS_CLR/Data/all 
knitr::opts_knit$set(root.dir = '~/Desktop/LMMS_CLR/Data/all')
```
Assign the directory as where ever you keep the biom files ".../Data/all".  

## Directory Assignment and Loading and merging the data 
```{r, warning=FALSE}
library(readr)
extractor <- function(i){ 
  #extracts raw count data from .biom files and names the genomic sequences as 16S_SEQUENCE
  
  temp <- as(biom_data(read_biom(i)),"matrix")
  temp <- cbind(rownames(temp), temp)
  rownames(temp) <- NULL 
  colnames(temp)[1]<-'16S_SEQUENCE'
  temp <- as.data.frame(temp)
  
  return(temp)
  }

#list of biom data
lst <- lapply(list.files(), FUN=extractor) 

#https://stackoverflow.com/questions/8091303/simultaneously-merge-multiple-data-frames-in-a-list
#https://stackoverflow.com/questions/48699066/r-merge-list-of-three-dataframes-into-single-dataframe-with-id-in-first-column 

#conducts full join and replaces NA non-recorded sequences as "0". 
x_merged <- lst %>% 
  purrr::reduce(dplyr::full_join, by = "16S_SEQUENCE") %>% 
  replace(., is.na(.), 0)

print(paste0("There are ", dim(x_merged)[1], " rows of uniquely recorded microbiota."))
```
## Changing the directory to where the meta data txt files are kept. 
```{r}
#change this line to where you downloaded the LMMS file. 
setwd("~/Desktop/LMMS_CLR/Data/tsvs")

#TSV are raw excel extracted metadata 
df_tsv <- list.files() %>% 
  lapply(function(i) {
    read.csv(i,sep = "\t")})
```
## Data Pre-processing
The reason for the varieties of \texttt{df_tsv}'s is for the purpose of debugging. 
```{r warning=FALSE}
#Preprocessing data types

#keeps are the variable names of the columns we want to keep 
keeps <- colnames(df_tsv[[7]]) 

#selects the columns with those names from keeps
df_tsv2 <- lapply(df_tsv, function(i){dplyr::select(i,keeps)})

#fixes annoying naming issues 
df_tsv2[[6]]$level_of_colonisation <- as.character(df_tsv2[[6]]$level_of_colonisation)
df_tsv2[[6]]$group_level_of_colonisation <- as.character(df_tsv2[[6]]$group_level_of_colonisation)

#Binding rows 
df_tsv3 <- df_tsv2 %>% bind_rows()

#creates a list of mice in the experiment 1 to 3 is due to which mice are known.
mouse_id<-lapply(LETTERS[c(1:6)],function(x){return(paste0("Mouse_",x,1:3))}) %>% unlist() 

print(mouse_id) 

#groups according to mouse_id then arranges by day (ascending)
df_tsv4 <- df_tsv3 %>% 
  filter(host_subject_id %in% mouse_id) %>% 
  group_by(host_subject_id) %>% 
  arrange(day_of_experiment_numerical, .by_group = TRUE)
```
Recall that \texttt{df_tsv} files are metadata files with \texttt{df_tsv3} and \texttt{df_tsv4} being the ones mostly used. 
## Filtering by continuous days (24 days was the result)
```{r}
lowest_day <- function(variables, df) {
  #Function to get the lowest amount of days in which there is continuous data across the board. 
  
  #INPUT: 
  # variables - mouse_id 
  # df - data frame in question
  
  #OUTPUT:
  # lowest day of continuous data
  
  full<-1:60 
  daylist <- df %>% dplyr::filter(host_subject_id == variables) %>% 
    ungroup() %>% 
    dplyr::select(day_of_experiment_numerical) %>% pull()
  smallest <- setdiff(full,daylist) %>% min() - 1
  
  return(smallest)
}

#does a global search for the lowest day of continuous data 
lowdays<-lapply(mouse_id,lowest_day, df=df_tsv4) %>% unlist() %>% min() 
lowdays 

#performs the "24" day cut off 
df_meta <- df_tsv4 %>% 
  dplyr::filter(day_of_experiment_numerical %in% 1:lowdays) 

#further cutting of the dataset 
df_present <- df_meta %>% dplyr::select(sample_id, title, antibiotic, 
                                         phase_of_experiment,
                                         level_of_colonisation,
                                         route_abx_administration,
                                         day_of_experiment_numerical, 
                                         average_cfu_counts, 
                                         cfu_counts)
```
## Acquiring sample id's 
Now that the meta data sets are correctly ordered we acquire the sample_id's in order to match the count data. 
```{r}
sample_E_name <- df_present %>% ungroup() %>% 
  dplyr::pull(sample_id) 
```
## OTU count filtering 
0.01 % cut across total OTU recorded. 
```{r}
percent = 0.01 
df_E <- x_merged %>% column_to_rownames(var="16S_SEQUENCE") %>% 
  mutate_all(., function(x) as.numeric(as.character(x))) %>% #converts the data frame to numeric 
  dplyr::select(all_of(sample_E_name)) %>%
  dplyr::filter(rowSums(.[,2:ncol(.)])*100/(sum(rowSums(.[,2:ncol(.)]))) > percent) %>% 
  dplyr::mutate_if(is.numeric, function(x) x+1)  %>% 
  tibble::rownames_to_column(., "Row.names")
```
## Taxa assignment DADA 
```{r}
#change the following line to /LMMS_CLR 
setwd("~/Desktop/Microbial Project/Reading 16s BIOM file 2")

#DNA sequences 
seq<-as.character(df_E$Row.names) 

#change the following line to where you will keep the dada2 file 
#something like /LMMS_CLR/Data/silva_nr99_v138.1_wSpecies_train_set.fa.gz should work. 

taxa_clasification <- assignTaxonomy(seq, "Data/silva_nr99_v138.1_wSpecies_train_set.fa.gz",
                                     multithread=TRUE)
write.csv(taxa_clasification, "Data/taxanomy3.csv")

#fixing annoying row.name issue. 
taxa_clasification2 <- cbind(Row.names = rownames(taxa_clasification), 
                             taxa_clasification)
rownames(taxa_clasification2) <- 1:nrow(taxa_clasification2)
df_taxa<-merge(taxa_clasification2, df_E, 
               by = "Row.names", all = TRUE)

#add OTUs labels
rownames(df_taxa) <- paste0("OTU", 1:nrow(df_taxa))
df_taxa <- df_taxa %>% add_column(OTU = paste0("OTU", 1:nrow(df_taxa)), .before = "Row.names")
```
## LMMS Processing Function 
The \texttt{multipurpose} function receives a list of mouse id's for one group of experiments and outputs a list containing various pieces of data (read comments for DATA). The most important result is \texttt{DATA$lmms.output} which is the lmms clr aggregated trajectories for the experimental group.  
```{r}
#multipurpose function 
multipurpose <- function(mouse_id_antibiotic, df_meta=df_meta, df_taxa=df_taxa, lowdays=lowdays) {
  #raw count data
  datacounts <- list() 
  
  #CLR data stored
  datalist <- list() 
  i <- 1 
  
  for (mouse in mouse_id_antibiotic) {
    print(mouse)
    #samples of mouse with respect to their day and "mouse" of the iteration 
    list_sample <- df_meta %>% filter(host_subject_id==mouse) %>% 
      arrange(day_of_experiment_numerical) %>% ungroup() %>% dplyr::select(sample_id)
    
    #gets the samples for mouse in loop 
    otumat<-df_taxa %>% dplyr::select(list_sample$sample_id) 
    
    #TAXA TABLE for all the mice 
    taxmat<-df_taxa %>% dplyr::select(colnames(taxa_clasification))
    
    OTU = otu_table(as.matrix(otumat), taxa_are_rows = TRUE) 
    TAX = tax_table(as.matrix(taxmat)) 
    
    #deals strictly in samples not OTUs 
    meta_sample <- df_meta %>% ungroup() %>% 
      dplyr::select(sample_id, phase_of_experiment, day_of_experiment_numerical) %>% 
      filter(sample_id %in% list_sample$sample_id) %>% arrange(day_of_experiment_numerical) 
    
    #fixes the indices
    meta_samplD2 <- meta_sample %>% dplyr::select(phase_of_experiment,day_of_experiment_numerical) %>%
      as.data.frame()
    rownames(meta_samplD2) <- meta_sample$sample_id
    
    physeq1 <- phyloseq(OTU,sample_data(meta_samplD2),TAX)
    physeq.top20 <- transform_sample_counts(physeq1, function(OTU) OTU/sum(OTU))
    
    #antibiotic name finder requires clean up 
    antibiotic <- df_meta %>% filter(host_subject_id==mouse) %>% ungroup() %>% 
      dplyr::select(antibiotic) %>% dplyr::pull() %>% dplyr::first()

    #plots the compositional data 
    print(plot_bar(physeq.top20, x="day_of_experiment_numerical", fill="Family") +
            facet_grid(~factor(phase_of_experiment, 
                               levels=c('naive','antibiotic','wean off','colonised','carrier','control')), 
                       scales="free_x") + 
            ggtitle(paste(mouse, antibiotic ,sep=" ")) +
            xlab("Day of Experiment"))
    
    #CLR transformed data 
    df_clr_D1 <- clr(t(otu_table(physeq1)[,1:lowdays]))
    rownames(df_clr_D1) <- paste0(mouse, "_", 1:lowdays)
    
    # stores the CLR transform inside datalist from before. 
    datalist[[i]] <- df_clr_D1 
    
    #original counts non-CLR
    df_clr_D2 <- t(otu_table(physeq1)[,1:lowdays])
    rownames(df_clr_D2) <- paste0(mouse, "_", 1:lowdays)
    
    # stores non-CLR transforms inside datacounts 
    datacounts[[i]] <- df_clr_D2 
    i<-i+1 
  }
  
  #CLR data, the following merges the data from list. 
  df_clr_ALL_E <- do.call(rbind, datalist)
  
  #CLR data merged from all mice in the experiment. 
  data.filtered <- df_clr_ALL_E
  time <- rep(1:lowdays, times=length(mouse_id_antibiotic), each=1)
  pred<-c(1:lowdays)
  
  #standard spline of step size 1.0 
  lmms.output <- lmms::lmmSpline(data = data.filtered, time = time,
                                 sampleID = rownames(data.filtered), deri = FALSE,
                                 basis = "p-spline", numCores = 6, timePredict = pred,
                                 keepModels = TRUE)
  
  #creates an interpolation spline of step size 0.5
  pred_fine<-seq(1,lowdays, by = 0.5)
  
  #standard spline of step size 0.5
  lmms.output_fine <- lmms::lmmSpline(data = data.filtered, time = time,
                                 sampleID = rownames(data.filtered), deri = FALSE,
                                 basis = "p-spline", numCores = 6, timePredict = pred_fine,
                                 keepModels = TRUE)
  
  #extracts spline out of the step size of 1 lmms 
  modelled.data <- t(slot(lmms.output, 'predSpline'))
  
  #tidification of the clr data
  data.gathered <- modelled.data %>% as.data.frame() %>% 
    tibble::rownames_to_column("time") %>%
    mutate(time = as.numeric(time)) %>%
    tidyr::pivot_longer(names_to="feature", values_to = 'value', -time)
  
  # print(ggplot(data.gathered, aes(x = time, y = value, color = feature)) + geom_line() +
  #   theme_bw() + ggtitle("`lmms` profiles") + ylab("Feature expression") +
  #   xlab("Time"))
  
  filter.res <- lmms.filter.lines(data = data.filtered, 
                                  lmms.obj = lmms.output, 
                                  time = time)
  
  profile.filtered <- filter.res$filtered[, colSums(filter.res$filtered != 0) > 0]
  
  #6 clusters
  vary_clusters <- 6
  pca.res <- pca(X = profile.filtered, ncomp = vary_clusters, scale = TRUE, center=TRUE) 

  #to store the data in a list
  DATA <- list() 
  DATA$time<- time #time step size 1 
  DATA$antibiotic <- antibiotic #antibiotics
  
  #clr data
  DATA$modelled.data <- modelled.data #lmms clr data of "3" mice from the same group 
  DATA$data.gathered <- data.gathered #tidy of lmms clr data for plotting purposes 
  DATA$data.filtered <- data.filtered #raw CLR data for 3 mice 
  
  DATA$pca.res <- pca.res #cluster assignment of OTUs 
  DATA$lmms.output <- lmms.output # interpolated LMMS data for step size 1 
  DATA$lmms.output_fine <- lmms.output_fine # interpolated LMMS data for step size 0.5  
  
  #raw count data
  DATA$datacounts <- datacounts
  
  return(DATA)
}
```
## LMMS CLR Data for control group 
```{r}
mice_names<-lapply(LETTERS[c(1:6)],function(x){return(paste0("Mouse_",x,1:3))})

mouse_control <- list(mice_names[[6]]) # index 6 to get the control experimental unit 

ctrl<-lapply(mouse_control, multipurpose, df_meta=df_meta, df_taxa=df_taxa, lowdays=lowdays) #apply the function 
```
\texttt{control_data} is the aggregated LMMS clr trajectory of the 3 three mice within the control group. Change the index of the \texttt{mice_names} above for the \texttt{mouse_control} variable to get other LMMS trajectories for other antibiotics. Below is a print out of the control trajectories. 
```{r}
ctrl[[1]]$antibiotic #gives the name of the experimental unit group  

control_data <- ctrl[[1]]$lmms.output@predSpline
head(control_data) 
```
```{r}
plotting_function <- function(clrlmms){
  # plots the clr splines from the main data source derived from the multipurpose function 
  
  p <- ggplot(clrlmms$data.gathered, aes(x = time, y = value, color = feature)) + geom_line() +
    theme_bw() + ggtitle(paste0(clrlmms$antibiotic," LMMS profiles"))  + ylab("Feature expression") +
    xlab("Time") + theme(legend.position = "none")
  return(p)} 

#plot of control lmms clr trajectories  
plotting_function(ctrl[[1]])
```
## Control Analysis of Models Used Tally 
```{r}
table(ctrl[[1]]$lmms.output@modelsUsed) 
```
## Control analysis
```{r}
mice_names<-lapply(LETTERS[c(1:6)],function(x){return(paste0("Mouse_",x,1:3))})
mouse_results<-lapply(mice_names, multipurpose, df_meta=df_meta, df_taxa=df_taxa, lowdays=lowdays) 
```
```{r}
table(mouse_results[[6]]$lmms.output@modelsUsed)
```

```{r}
control_analysis <- function(clrlmms, exp=5,ctrl=6){ 
  # Displays a plot of the respective splines for the OTUs of each control unit. 
  # Since there are many OTUs, the 20th to 30th OTUs are selected out of the filter. 
  # modify 20 to 30 index otu_fil <- otu_filtered$Molecule[20:30].  
  
  # INPUTS: 
  # clrlmms - clr data frame 
  # ctrl=6 index of the control experimental unit 
  # i=5 index of the vancomycin experimental unit. Change index to test out different experimental units.
  
  # OUTPUTS: 
  # plot of clr LMMS splines with raw clr data points 
  
  #times for the differential analysis
  df_times <- c(clrlmms[[exp]]$time, clrlmms[[ctrl]]$time)
  
  #raw non-lmms raw clr data for each mouse
  df_all_control_F<- rbind(clrlmms[[exp]]$data.filtered, clrlmms[[ctrl]]$data.filtered) 
  
  #data point names for each mouse and time point 
  df_de_names <- row.names(df_all_control_F)
  
  #assigns the labels for each experimental group  
  df_group <- c(replicate(length(clrlmms[[exp]]$time),"G1"), 
                replicate(length(clrlmms[[ctrl]]$time),"G2"))   
  
  #lmmsDE analysis
  lmmsDEtest <-lmmsDE(data=df_all_control_F,
                        time=df_times,
                        sampleID=df_de_names,
                        group=df_group)
  #for debugging purposes
  lmmsDEtest2<-lmmsDEtest
  
  myvars <- paste("G1", clrlmms[[exp]]$time, sep=" ")
  myvars2 <- paste("G2",clrlmms[[ctrl]]$time, sep=" ") 
  
  otu_filtered <- lmmsDEtest2@DE %>% filter(adj.Group<0.05) %>% dplyr::select(Molecule) %>% 
    lapply(function(x) as.numeric(gsub("OTU", "", x))) 
  
  #takes the middle 10 from 20 to 30.
  otu_fil <- otu_filtered$Molecule[20:30]
  
  group_1 <- lmmsDEtest2@predGroup[,myvars][otu_fil,] %>% as.data.frame() %>%
    add_column(OTUs = otu_fil, .before = 1) %>%
    add_column(Group = rep("Experimental", times=length(otu_fil)), .before = 2) %>% 
    pivot_longer(., -c("Group", "OTUs"), values_to = "Value", names_to = "Time")

  group_2 <-lmmsDEtest2@predGroup[,myvars2][otu_fil,] %>% as.data.frame() %>% 
    add_column(OTUs = otu_fil, .before = 1) %>% 
    add_column(Group = rep("Control", times=length(otu_fil)), .before = 2) %>% 
    pivot_longer(., -c("Group", "OTUs"), values_to = "Value", names_to = "Time")
  
  super_frame <- rbind(group_1,group_2) 
  super_frame$Time <- rep(ult[[ctrl]]$time, times=2*length(otu_fil))
  
  col_bind <- rbind(clrlmms[[exp]]$data.filtered, clrlmms[[ctrl]]$data.filtered) %>% 
    as_tibble(.,rownames = "mouse_id") %>% 
    add_column(Group = c(rep("Experimental", 
                         times=length(clrlmms[[exp]]$time)), 
                        rep("Control", 
                         times=length(clrlmms[[ctrl]]$time))), .before = 1) %>% 
    add_column(Time = c(clrlmms[[exp]]$time, clrlmms[[ctrl]]$time), .before = 1)
  
  other_frame <- col_bind %>% dplyr::select(-c("mouse_id")) %>%  
    pivot_longer(-c("Group", "Time"), values_to = "Value", names_to = "OTUs") %>% 
    mutate(OTUs = as.numeric(gsub("OTU", "", OTUs))) %>% 
    filter(OTUs %in% otu_fil)  

  p <- ggplot() +
        geom_line(data = super_frame, aes(x = Time, y = Value, color = Group)) + 
        geom_point(data = other_frame, aes(x = Time, y = Value, color = Group)) +
        facet_wrap(facets =  vars(OTUs)) + 
        ggtitle(paste0("OTUs of Group Difference LMMSDE: ", 
                       clrlmms[[exp]]$antibiotic," and ", 
                       clrlmms[[ctrl]]$antibiotic ))  
  return(p)
}

control_analysis(mouse_results, exp=5, ctrl=6)
```
```{r}

library(igraph) #network plotting 
library(gridExtra) #grid.arrange function

plot_networks <- function(clrlmms, exp=5, ctrl=6, proportion=FALSE){ 
  # provides network depiction of the correlation structure of the microbiome
  
  # INPUTS: 
  #exp=5 # - antibiotic index in question.
  #ctrl=6 # - control unit index. 
  
  # OUTPUTS: 
  # two network plots the first being the control, second being antibiotic. 
  
  #PC1 postive OTUs for antibiotic
  OTU_PC1_POS <- which(data.frame(clrlmms[[exp]]$pca.res$loadings$X)>0, arr.ind = T) %>% 
    data.frame() %>% dplyr::filter(col==1) %>% 
    row.names() %>% gsub("\\..*","",.) %>% 
    stri_unique() 
  
  #PC1 negative OTUs for antibiotic
  OTU_PC1_NEG <- which(data.frame(clrlmms[[exp]]$pca.res$loadings$X)<0, arr.ind = T) %>% 
    data.frame() %>% dplyr::filter(col==1) %>% 
    row.names() %>% gsub("\\..*","",.) %>% 
    stri_unique()
  
  #ALL PC1 OTUs for antibiotic
  PC1_component <- c(OTU_PC1_POS, OTU_PC1_NEG) %>% unique()
  
  #df1 collects 
  #recall that $data.gathered is the tidified version of clr lmms splines. 
  df_1 <- clrlmms[[exp]]$data.gathered %>% group_by(feature) %>% 
    mutate(value=scale(value)) %>% 
    filter(feature %in% OTU_PC1_POS)
  
  plot1 <- ggplot(df_1, aes(x = time, y = value[,1], color = feature)) + geom_line() +
    theme_bw() + ggtitle(paste0(clrlmms[[exp]]$antibiotic," PC1 positive lmms profiles")) + 
    ylab("Feature expression") +
    xlab("Time") +
    theme(legend.position = "none")
  
  df2 <- clrlmms[[exp]]$data.gathered %>% group_by(feature) %>% 
    mutate(value=scale(value)) %>% 
    filter(feature %in% OTU_PC1_NEG)
  
  plot2 <- ggplot(df2, aes(x = time, y = value[,1], color = feature)) + geom_line() +
    theme_bw() + ggtitle(paste0(clrlmms[[exp]]$antibiotic," PC1 negative lmms profiles")) + 
    ylab("Feature expression") +
    xlab("Time") +
    theme(legend.position = "none")
  
  grid.arrange(plot1, plot2, ncol=2) %>% show()
  
  
  #CONTROL DATAFRAME
  #dataframe to store information about PC1 component of control 
  df_control <- list()
  
  #lmms unscaled and uncentred for control data for PC1 component OTUs 
  df_control$clr <- clrlmms[[ctrl]]$modelled.data %>% data.frame() %>%                          
    dplyr::select(all_of(PC1_component))
  
  df_control$anti <- clrlmms[[ctrl]]$antibiotic #acquires antibiotic
  
  
  #ANTIBIOTIC DATAFRAME
  #dataframe to store information about PC1 component of control 
  df_antibiotic <- list()
  
  #lmms unscaled or centred for antibiotic data for PC1 component OTUs 
  df_antibiotic$clr <- clrlmms[[exp]]$modelled.data %>% data.frame() %>%
    dplyr::select(all_of(PC1_component))
  
  df_antibiotic$anti <- clrlmms[[exp]]$antibiotic #acquires antibiotic 
  
  #stores dataframes in list antibiotic first, control second 
  vector <- list(df_antibiotic, df_control)
  par(mfrow=c(1,2))  
  for (i in c(1,2)) {
    
    set.seed(573313)
    df<-vector[[i]]$clr
    
    #correlation calculator 
    a <- cor(df, y = NULL, use = "all.obs", method = "spearman") #%>% as.matrix()
    a[abs(a) < 0.9] <- 0
    diag(a) <- 0 # remove loops.
    # create igraph object.
    
    if (proportion==TRUE) {
      #something weird going on with the proportionality matrix, spca removes OTUs making it 
      #hard to compare 
      a<-clrlmms[[i]]$proportion$propr.distance[all_of(PC1_component),all_of(PC1_component)]
      a <- as.matrix(a)
      diag(a) <- 0
      print(dim(a))
    }
    
    #enlist all nodes of color to the antibiotic in question 
    PC1_pos<-OTU_PC1_POS
    PC1_neg<-OTU_PC1_NEG

    g <- graph.adjacency(a, weighted=TRUE)
    
    #
    #delete.edges(g, which(abs(E(g)$weight)<0.97))
    #print(E(g)$width)
    # assign edge's width as a function of weights.
    #E(g)$width <- 2#2*exp(abs(E(g)$weight))
    #print(E(g)$weight)
    
    alpha<- 0.4
    E(g)$color<-ifelse(a>0, adjustcolor("red",alpha.f = alpha),
                        adjustcolor("blue",alpha.f = alpha))
    
    V(g)[PC1_pos]$color <- adjustcolor("green",alpha.f = alpha)
    V(g)[PC1_neg]$color <- adjustcolor("orange",alpha.f = alpha)
    
    wrap_strings <- function(vector_of_strings,width){
      as.character(sapply(vector_of_strings, FUN=function(x){
        paste(strwrap(x, width=width), collapse="\n")
      }))
    }
    
    # Apply the function to wrap the node labels
    #V(g)$label = wrap_strings(V(g)$label, 12)
    
    ## Shrink font
    V(g)$label.cex = 0.6
    #print("vert count", vcount(g))  
    
    # Function to increase node separation (for explanatory details, see the link below)
    # Source: http://stackoverflow.com/a/28722680/496488
    layout.by.attr <- function(graph, wc, cluster.strength=1, layout=layout.auto) {  
      g <- graph # create a lightweight copy of graph w/o the attributes.
      E(g)$weight <- 10
      
      attr <- cbind(id=1:vcount(g), val=wc)
      g <- g + vertices(unique(attr[,2])) + igraph::edges(unlist(t(attr)), weight=cluster.strength)
      #print(vcount(graph))
      #print(layout(g, weights=E(g)$weight))
      l <- layout(g, weights=E(g)$weight)[1:(vcount(graph)),]
      return(l)
    }
    
    plot(g, edge.width = E(g)$width, 
         main= vector[[i]]$anti,
         mode = "undirected",
         vertex.frame.color = NA, 
         vertex.size = 15, 
         vertex.label=NA, 
         layout=layout.by.attr(g, wc=4, cluster.strength = 3), 
         edge.arrow.size = 0.0)
  }
}
plot_networks(mouse_results)

```

